cranelift/filetests/filetests/interpreter/iaddcarry.clif

```diff
@@ -0,0 +1,73 @@
+test interpret
```

Comment:
It looks like these instructions don't work at all in any of the backends. This is slightly concerning to me because I suspect that they may be legacy backend only instructions which will be deprecated soon. (is this the case @cfallin?).

That being said, and assuming that they are a TODO item on the backends, I think we should move these tests to the `runtests` folder, even if they only run in the interpreter. Reason being, that eventually we will implement them, and these tests are a lot easier to find there. 

One of the things that I've been working towards is removing the `interpreter` folder and moving all test cases to the `runtests` folder. I think that the distinction isn't very meaningful. We already have some [tests](https://github.com/bytecodealliance/wasmtime/blob/a662f5361de2edeaa1b0ba8fdcc16226eb827453/cranelift/filetests/filetests/runtests/stack-addr-32.clif) there that only work in the interpreter, but we never explicitly discussed this. Thoughts @cfallin, @abrown  ?

Reply:
I dug into the history of these instructions: `iadd_carry`, `iadd_cin`,  and `iadd_cout` are not used by the Wasm-to-CLIF converter in `code_translator.rs` and do not appear in any of the ISA-specific lowering code in the new backend (e.g., `cranelift/codegen/src/isa/*`). They were added by @ryzokuken almost two years ago  in https://github.com/bytecodealliance/cranelift/pull/1005 and, unless they or @bjorn3  are still using these instructions, I propose we remove them. (@afonso360, I'm fine with moving stuff into the `runtests` folder).

Reply:
The problem with this is that I hate to get rid of good code: thanks @dheaton-arm and @afonso360 for fleshing out the interpreter! We've talked before about cleaning up the CLIF opcode space and that probably should have happened before we jumped in on the interpreter, sorry; @cfallin can correct me if I'm wrong but I would think any CLIF opcode [here](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/codegen/src/isa/aarch64/lower_inst.rs#L3315-L3342) or anything starting with `X86...` is a likely candidate for removal?

Reply:
I currently don't use them as backend support is bad, but if backend support is implemented I will want to use at least the `_cout` variants to detect overflows efficiently.

https://github.com/bjorn3/rustc_codegen_cranelift/blob/9f5b52045c928157b12bec1d670e220fc7597375/src/num.rs#L198-L225

The `_carry` and I think `_cin` variants are necessary to efficiently implement certain llvm intrinsics used by stably exported simd intrinsics:

https://github.com/bjorn3/rustc_codegen_cranelift/blob/9f5b52045c928157b12bec1d670e220fc7597375/src/intrinsics/llvm.rs#L145-L181

Reply:
> @cfallin can correct me if I'm wrong but I would think any CLIF opcode here or anything starting with X86... is a likely candidate for removal?

The `_imm` opcodes are currently legalized to non-`_imm` variants. I use the `_imm` variants extensively in cg_clif as they are much more readable both in the cg_clif source and in textual clif ir form. The `if`/`ff` variants of instructions should be removed though IMO. Just like the `x86_` instructions.

Reply:
I guess the decision should be framed as an "all or nothing:" either we remove them from the interpreter and CLIF or we implement them in the interpreter AND the backends. E.g., the `_imm` variants are implemented in the interpreter and it wouldn't be too difficult to add support in the backends.